cmd/nsm-sriov-manager/nsm-sriov-manager.go

```diff
@@ -0,0 +1,418 @@
+				err, strings.Replace(vf.pciAddr, "-", ":", -1))
+			delete(discoveredVFs.vfs, vf.pciAddr)
+			continue
+		}
+	}
+	return nil
+}
+
+func waitAndRetry(timeout time.Duration, retries int, check func() bool) error {
+	ticker := time.NewTicker(timeout)
+	for r := 0; r < retries; r++ {
+		select {
+		case <-ticker.C:
+			if check() {
+				return nil
+			}
+		}
+
+	}
+
+	return fmt.Errorf("timeout has expired")
+}
+
+func bindVF(vf *VF) error {
+	pciAddr := strings.Replace(vf.pciAddr, "-", ":", -1)
+	unbindPath := fmt.Sprintf("/sys/bus/pci/devices/%s/driver/unbind", pciAddr)
+	cmdUnbind := exec.Command("echo", pciAddr)
+	u, err := os.OpenFile(unbindPath, os.O_WRONLY, 0200)
+	if err != nil {
+		return fmt.Errorf("fail to open unbind path %s with error: %+v", unbindPath, err)
+	}
+	defer u.Close()
+	cmdUnbind.Stdout = u
+	if err := cmdUnbind.Run(); err != nil {
+		return fmt.Errorf("unbind failed with error: %+v", err)
+	}
+
+	bindArgs := fmt.Sprintf(" %s %s ", vf.pciVendor, vf.pciType)
+	cmdBind := exec.Command("echo", bindArgs)
+	bindPath := "/sys/bus/pci/drivers/vfio-pci/new_id"
+	b, err := os.OpenFile(bindPath, os.O_WRONLY, 0200)
+	if err != nil {
+		return fmt.Errorf("fail to open bind path %s with error: %+v", unbindPath, err)
+	}
+	defer b.Close()
+	cmdBind.Stdout = b
+	if err := cmdBind.Run(); err != nil {
+		return fmt.Errorf("bind failed with error: %+v", err)
+	}
+
+	return waitAndRetry(200*time.Millisecond, 5, func() bool {
+		_, err := os.Stat(vf.VFIODevice)
+		if err == nil {
+			return true
+		}
+		return false
+	})
+}
+
+func main() {
+	flag.Set("logtostderr", "true")
+	flag.Parse()
+	logrus.Infof("Starting SRIOV Network Device Plugin...")
+	discoveredVFs := newVFs()
+	if err := discoverNetworks(discoveredVFs); err != nil {
+		logrus.Errorf("%+v", err)
+		os.Exit(1)
+	}
+
+	// Uncomment to mock VFs for unit testing
+	// discoveredVFs := mockVFs()
+
+	if len(discoveredVFs.vfs) == 0 {
+		logrus.Info("No VF were discovered, exiting...")
+		os.Exit(0)
+	}
+
+	logrus.Infof("%d VFs were discovered on the host.", len(discoveredVFs.vfs))
+	// Building vfio device for each VF
+	if err := buildVFIODevices(discoveredVFs); err != nil {
+		logrus.Errorf("Failed to build VFIO devices for VFs with error: %+v", err)
+		os.Exit(1)
+	}
+	cf, err := buildSRIOVConfigMap(discoveredVFs)
+	if err != nil {
+		logrus.Errorf("Failed to build SRIOV config map with error: %+v", err)
+		os.Exit(1)
+	}
+
+	configMap, err := yaml.Marshal(cf)
+	if err != nil {
+		logrus.Errorf("Failed to marshal SRIOV config map with error: %+v", err)
+		os.Exit(1)
+	}
+
+	if err := ioutil.WriteFile("nsm-sriov-configmap.yaml", configMap, 0644); err != nil {
+		logrus.Errorf("Failed to save  SRIOV config map with error: %+v", err)
+		os.Exit(1)
+	}
+	logrus.Info("sriov configmap for Network service mesh has been saved in nsm-sriov-configmap.yaml")
+}
+
+// mockVFs will be used for unit testing
+func mockVFs() *VFs {
```

Comment:
If it isn't used for testing already then it probably should be in a subsequent commit

Reply:
I was using it for testing the rest of code on my mac. I will create a separate _test.go file and put it there.